pkg/operator/controllers/pullsecret/pullsecret_controller.go

```diff
@@ -162,3 +135,90 @@ func (r *PullSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Named(controllers.PullSecretControllerName).
 		Complete(r)
 }
+
+// fixAndUpdateGlobalPullSecret checks the state of the pull secrets, in case of missing or broken ARO pull secret
+// it replaces it with working one from controller Secret
+func (r *PullSecretReconciler) fixAndUpdateGlobalPullSecret(ctx context.Context, operatorSecret, userSecret *corev1.Secret) (secret *corev1.Secret, err error) {
+	if operatorSecret == nil {
+		return nil, errors.New("nil operator secret, cannot verify userData integrity")
+	}
+
+	create := userSecret == nil
+	remove := false
+
+	// userSecret can happen to have broken type, which means it have to be recreated
+	// with proper type
+	if userSecret != nil && (userSecret.Type != corev1.SecretTypeDockerConfigJson || userSecret.Data == nil) {
+		create = true
+		remove = true
+	}
+
+	if create {
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      pullSecretName.Name,
+				Namespace: pullSecretName.Namespace,
+			},
+			Type: corev1.SecretTypeDockerConfigJson,
+		}
+		secret.Data = make(map[string][]byte)
+	} else {
+		secret = userSecret.DeepCopy()
+		if !json.Valid(secret.Data[corev1.DockerConfigJsonKey]) {
+			delete(secret.Data, corev1.DockerConfigJsonKey)
+		}
+	}
+
+	fixedData, update, err := pullsecret.Merge(string(secret.Data[corev1.DockerConfigJsonKey]), string(operatorSecret.Data[corev1.DockerConfigJsonKey]))
+	if err != nil {
+		return nil, err
+	}
+
+	if !create && !update {
+		return userSecret, nil
+	}
+
+	if remove {
+		// unfortunately the type field is immutable.
+		err := r.kubernetescli.CoreV1().Secrets(secret.Namespace).Delete(ctx, secret.Name, metav1.DeleteOptions{})
+		if err != nil && !kerrors.IsNotFound(err) {
+			return nil, err
+		}
+	}
+
+	secret.Data[corev1.DockerConfigJsonKey] = []byte(fixedData)
+
+	if create {
+		secret, err := r.kubernetescli.CoreV1().Secrets(secret.Namespace).Create(ctx, secret, metav1.CreateOptions{})
+		return secret, err
+	}
+
+	secret, err = r.kubernetescli.CoreV1().Secrets(secret.Namespace).Update(ctx, secret, metav1.UpdateOptions{})
+	return secret, err
+}
+
+// parseRedHatKeys unmarshal and extract following RH keys from pull-secret:
+//   - redhat.registry.io
+//   - cloud.redhat.com
+//   - registry.connect.redhat.com
+// if present, return error when the parsing fail, which means broken secret
+func (r *PullSecretReconciler) parseRedHatKeys(secret *corev1.Secret) (foundKeys []string, err error) {
+	// parse keys and validate JSON
+	parsedKeys, err := pullsecret.UnmarshalSecretData(secret)
+	if err != nil {
+		r.log.Info("pull secret is not valid json - recreating")
+		return foundKeys, err
+	}
+
+	if parsedKeys == nil {
+		return foundKeys, nil
+	}
+
+	for _, rhKey := range rhKeys {
+		if v := parsedKeys[rhKey]; len(v) > 0 {
+			foundKeys = append(foundKeys, rhKey)
+		}
+	}
+
+	return foundKeys, nil
```

Comment:
We're returning `foundKeys, nil` here regardless, so this could be simplified:

```suggestion
	if parsedKeys != nil {
		for _, rhKey := range rhKeys {
			if v := parsedKeys[rhKey]; len(v) > 0 {
				foundKeys = append(foundKeys, rhKey)
			}
		}
	}
	return foundKeys, nil
```

Reply:
ah good eye, thanks